libyul/optimiser/CircularReferencesPruner.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libyul/optimiser/CircularReferencesPruner.h>
+
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/OptimizerUtilities.h>
+#include <libyul/AsmData.h>
+
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+
+using namespace std;
+using namespace solidity::yul;
+
+void CircularReferencesPruner::run(OptimiserStepContext& _context, Block& _ast)
+{
+	CircularReferencesPruner{_context.reservedIdentifiers}(_ast);
+}
+
+void CircularReferencesPruner::operator()(Block& _block)
+{
+	ASTModifier::operator()(_block);
+
+	set<YulString> functionsToKeep =
+		functionsCalledFromOutermostContext(CallGraphGenerator::callGraph(_block));
+
+	for (auto&& statement: _block.statements)
+		if (holds_alternative<FunctionDefinition>(statement))
+		{
+			FunctionDefinition const& funDef = std::get<FunctionDefinition>(statement);
+			if (!functionsToKeep.count(funDef.name))
+				statement = Block{};
+		}
+
+	removeEmptyBlocks(_block);
+}
+
+set<YulString> CircularReferencesPruner::functionsCalledFromOutermostContext(CallGraph const& _callGraph)
+{
+	set<YulString> verticesToTraverse(util::convertContainer<set<YulString>>(m_reservedIdentifiers));
```

Comment:
now you don't need `convertContainer` anymore, because the types are the same, right?

Reply:
Right, fixed here https://github.com/ethereum/solidity/pull/8267/commits/d757eed2a8728f7c61675c849b04315eb00da490